### ISSUE TITLE
Better normalize input CLIP embeddings when training prior

### ISSUE THREAD

**Issue by crowsonkb** [lucidrains/DALLE2-pytorch#60]
Right now it looks like the CLIP embeddings (text to condition on and the clean image embedding to train the prior to predict) are being normalized to norm 1 when training the prior. This is very badly scaled for Gaussian diffusion and I suspect it is a reason why LAION's attempt at training a prior is going badly. With a 512 dim CLIP embedding, normalizing to norm 1 means that the values will be around the range -0.06 to 0.06 or some such. I suggest either finding the elementwise mean and std of the CLIP image embeddings in the training set and normalizing them to mean 0 std 1, or using PCA, or (what I am doing with my prior) simply normalizing the embeddings then multiplying them by sqrt(embed_dim), to make them better scaled for Gaussian diffusion.

This scaling makes the embeddings match the expected norm of the Gaussian noise. When we combines the clean embedding and the noise, we do so according to alpha^2 + sigma^2 = 1, where alpha and sigma are the scaling factors for the clean embedding and noise at that timestep. Since the noise is nearly always nearly orthogonal to the embedding the forward process paths stay near the manifold (the hypersphere w/ the radius sqrt(embed_dim).) I did this because it was easier than figuring out diffusion on Riemannian manifolds.

**Comment by lucidrains** [comment]
@crowsonkb ohh got it, thanks for the advice Katherine! i'll make the `sqrt(embed_dim)` an option tomorrow morning

**Comment by lucidrains** [comment]
@crowsonkb btw, i saw you using Assaf's ResizeRight! i was just evaluating that a few days ago

decided to release it as a pypi package since I saw you using it successfully https://pypi.org/project/resize-right/

**Comment by crowsonkb** [comment]
Oh, thank you for making it a package on pypi! I was going to do that eventually at some point but never got around to it. :)

**Comment by XinYu-Andy** [comment]
> 

Hi,
Thanks for your helpful information, I am wondering what do you mean about "This scaling makes the embeddings match the expected norm of the Gaussian noise"?
Do you mean if we normalize each channel across the dataset such that its std=1, then we can keep its std to be 1 when adding some noise to it? If so, why can this operation help the training?
